src/sudokulogic.hs

```diff
@@ -0,0 +1,307 @@
+module SudokuLogic where
+
+import Sudoku
+import Sudokus
+import Data.List
+import Control.Applicative (Alternative(empty))
+import Debug.Trace (trace)
+import Data.Type.Coercion (trans)
+import Data.Maybe
+--import Test.QuickCheck hiding ((===))
+
+
+type Lemma = Sudoku -> Position -> Bool
+
+-- | A sorted list of all the lemmas avalible sorted in a ~smart~ order	 
+lemmas :: [Lemma]
+lemmas = [lemmaLastCellInBlock,
+          lemmaLastCellInColumn,
+          lemmaLastCellInBlock]
+
+
+-- | A lemma that checks if a value is the last in a block
+lemmaLastCellInBlock :: Sudoku -> Position -> Bool
+lemmaLastCellInBlock s pos = lastFreeCellInSection s pos (blockFromPos s pos)
+
+-- | A lemma that checks if a value is the last in a row
+lemmaLastCellInRow :: Sudoku -> Position -> Bool
+lemmaLastCellInRow s pos = lastFreeCellInSection s pos (rowFromPos s pos)
+
+-- | A lemma that checks if a value is the last in a column
+lemmaLastCellInColumn :: Sudoku -> Position -> Bool
+lemmaLastCellInColumn s pos = lastFreeCellInSection s pos (colFromPos s pos)
```

Comment:
Instead of having methods that check all the time whether these techniques are applicable, you could have some data structures in the state of your solver from which you can infer where these techniques can be used.

For example:
* for each line, a number for how many values are still missing
* for each cell, how many candidates it has

These could also be ordered by these numbers so that you e.g. don't have to go over all cells to figure out which one has the least candidates remaining.